File: iosApp/iosApp/Modules/KeyStore ✅/KeyStoreWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - DefaultKeyStoreWireframe

final class DefaultKeyStoreWireframe {

    private weak var parent: UIViewController?
    private let keyStoreService: KeyStoreService
    private let networksService: NetworksService
    private let newMnemonic: MnemonicNewWireframeFactory
    private let updateMnemonic: MnemonicUpdateWireframeFactory
    private let importMnemonic: MnemonicImportWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory

    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        keyStoreService: KeyStoreService,
        networksService: NetworksService,
        newMnemonic: MnemonicNewWireframeFactory,
        updateMnemonic: MnemonicUpdateWireframeFactory,
        importMnemonic: MnemonicImportWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory
    ) {
        self.parent = parent
        self.keyStoreService = keyStoreService
        self.networksService = networksService
        self.newMnemonic = newMnemonic
        self.updateMnemonic = updateMnemonic
        self.importMnemonic = importMnemonic
        self.alertWireframeFactory = alertWireframeFactory
    }
}

// MARK: - KeyStoreWireframe

extension DefaultKeyStoreWireframe {

    func present() {
        let vc = wireUp()
        self.vc = vc
        parent?.asEdgeCardsController?.setBottomCard(vc: vc) ?? parent?.show(vc, sender: self)
    }

    func navigate(with destination: KeyStoreWireframeDestination) {
        let edgeVc = parent as? EdgeCardsController
        if destination is KeyStoreWireframeDestination.HideNetworksAndDashboard {
            edgeVc?.setDisplayMode(.bottomCard)
        }
        if destination is KeyStoreWireframeDestination.Networks {
            edgeVc?.setDisplayMode(.overviewTopCard, animated: true)
        }
        if destination is KeyStoreWireframeDestination.Dashboard {
            edgeVc?.setDisplayMode(.master, animated: true)
        }
        if destination is KeyStoreWireframeDestination.DashboardOnboarding {
            edgeVc?.setDisplayMode(.masterOnboardAnim, animated: true)
        }
        if let input = destination as? KeyStoreWireframeDestination.NewMnemonic {
            let context = MnemonicNewWireframeContext(handler: input.handler)
            newMnemonic.make(vc, context: context).present()
        }
        if let input = destination as? KeyStoreWireframeDestination.ImportMnemonic {
            let context = MnemonicImportWireframeContext(handler: input.handler)
            importMnemonic.make(vc, context: context).present()
        }
        if let input = destination as? KeyStoreWireframeDestination.EditKeyStoreItem {
            let context = MnemonicUpdateWireframeContext(
                keyStoreItem: input.item,
                onUpdateHandler: input.handler,
                onDeleteHandler: input.onDeleted
            )
            updateMnemonic.make(vc, context: context).present()
        }
        if destination is KeyStoreWireframeDestination.ConnectHardwareWallet {
            alertWireframeFactory.make(vc, context: .underConstructionAlert()).present()
        }
        if destination is KeyStoreWireframeDestination.ImportPrivateKey {
            alertWireframeFactory.make(vc, context: .underConstructionAlert()).present()
        }
        if destination is KeyStoreWireframeDestination.CreateMultisig {
            alertWireframeFactory.make(vc, context: .underConstructionAlert()).present()
        }
    }
}

private extension DefaultKeyStoreWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultKeyStoreInteractor(
            keyStoreService: keyStoreService,
            networksService: networksService
        )
        let vc: KeyStoreViewController = UIStoryboard(.keyStore).instantiate()
        let presenter = DefaultKeyStorePresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )

        vc.presenter = presenter
        return NavigationController(rootViewController: vc)
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/keystore/KeyStoreWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.keystore

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.R
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicimport.MnemonicImportWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicnew.MnemonicNewWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.keyStore.DefaultKeyStoreInteractor
import com.sonsofcrypto.web3walletcore.modules.keyStore.DefaultKeyStorePresenter
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreWireframe
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.mnemonicImport.MnemonicImportWireframeContext
import com.sonsofcrypto.web3walletcore.modules.mnemonicNew.MnemonicNewWireframeContext

class DefaultKeyStoreWireframe(
    private val parent: WeakRef<Fragment>?,
    private val keyStoreService: KeyStoreService,
    private val networksService: NetworksService,
    private val mnemonicNewWireframeFactory: MnemonicNewWireframeFactory,
    private val mnemonicImportWireframeFactory: MnemonicImportWireframeFactory,
): KeyStoreWireframe {

    private lateinit var fragment: WeakRef<Fragment>

    override fun present() {
        val fragment = wireUp()
        this.fragment = WeakRef(fragment)
        parent?.get()?.childFragmentManager?.beginTransaction()?.apply {
            add(R.id.container, fragment)
            commitNow()
        }
    }

    override fun navigate(destination: KeyStoreWireframeDestination) {
        when (destination) {
            is KeyStoreWireframeDestination.NewMnemonic -> {
                val context = MnemonicNewWireframeContext(destination.handler)
                mnemonicNewWireframeFactory.make(fragment?.get(), context).present()
            }
            is KeyStoreWireframeDestination.ImportMnemonic -> {
                val context = MnemonicImportWireframeContext(destination.handler)
                mnemonicImportWireframeFactory.make(fragment?.get(), context).present()
            }
            else -> { println("[AA] handle event $destination") }
        }
    }

    private fun wireUp(): Fragment {
        val view = KeyStoreFragment()
        val interactor = DefaultKeyStoreInteractor(
            keyStoreService,
            networksService,
        )
        val presenter = DefaultKeyStorePresenter(
            WeakRef(view),
            this,
            interactor,
        )
        view.presenter = presenter
        return NavigationFragment(view)
    }
}